04-onion-routing.md

```diff
@@ -432,32 +435,143 @@ otherwise meets the amount criterion (eg. some other failure, or
+
+Nodes receiving onion packets (either onion messages, or payments) may
+hide their identity from senders by providing a `blinded_path`.  To do
+this the recipient needs to hide both the real node ids, and the
+instructions to those nodes on where to forward the onion.
+
+1. subtype: `blinded_path`
+2. data:
+   * [`point`:`first_node_id`]
+   * [`point`:`blinding`]
+   * [`byte`:`num_hops`]
+   * [`num_hops*onionmsg_hop`:`path`]
+
+1. subtype: `onionmsg_hop`
+2. data:
+    * [`point`:`blinded_node_id`]
+    * [`u16`:`enclen`]
+    * [`enclen*byte`:`encrypted_recipient_data`]
+
+A blinded path consists of:
+1. an initial introduction point (`first_node_id`)
+2. an initial tweak to modify the first node_id to decrypt the onion (`blinding`)
+3. a series of tweaked node ids (`path.blinded_node_id`)
+4. a series binary blobs encrypted to the real node ids (`path.encrypted_recipient_data`)
+   to tell them the next hop.
+
+For example, Dave wants Alice to reach him via public node Bob then
+Carol.  He creates and encrypts three `encrypted_data_tlv`s:
+1. blob_bob: For Bob to tell him to forward to Carol
+2. blob_carol: For Carol to tell her to forward to him
+3. blob_dave: For himself to indicate the path was used, and any metadata he wants.
+
+To mask the node ids, he derives three blinding factors, which turn
+Bob into Bob', Carol into Carol' and Dave into Dave'.  These are a simple
+chain, so Bob can derive Carols without having to be told explicitly.
+
+So this is the `blinded_path` he hands to Alice.
+
+1. `first_node_id`: Bob
+2. `blinding`: to turn Bob into Bob'
+3. `path`: [Bob', bob_blob], [Carol', carol_blob], [Dave', dave_blob]
+
+Alice encrypts an onion to Bob', Carol', Dave' and gives it to Bob
+with the first blinding factor `blinding`.
+
+Bob uses the blinding and his private key to decrypt the first layer
+of the onion (created by Alice), and uses his normal private key to
+decrypt "bob_blob" (created by Dave).  The blob decrypts into a
+`encrypted_data_tlv` which indicates where the onion is to be
+forwarded (i.e. Carol).  Bob derives the next `blinding` and sends it
+an the onion to Carol.
+
+### Requirements
+
+Note that the creator of the blinded path (i.e. the recipient) is creating it for the sender to use to create an onion, and for the intermediate nodes to read the instructions, hence there are two reader sections here.
+
+The writer of `blinded_path`:
+
+- MUST create a viable path to itself ($`N_r`$) i.e. $`N_0 \rightarrow N_1 \rightarrow ... \rightarrow N_r`$.
+- MUST set `first_node_id` to $`N_0`$
+- MUST create a series of ECDH shares secrets for each node in the route using the following algorithm:
+  - $`e_0 /leftarrow {0;1}^256`$ ($`e_0`$ SHOULD be obtained via CSPRG)
+  - $`E_0 = e_0 \cdot G`$
+  - For every node in the route:
+    - let $`N_i = k_i * G`$ be the `node_id` ($`k_i`$ is $`N_i`$'s private key)
+    - $`ss_i = SHA256(e_i * N_i) = SHA256(k_i * E_i)$` (ECDH shared secret known only by $`N_r`$ and $`N_i`$)
```

Comment:
```suggestion
    - $ss_i = SHA256(e_i * N_i) = SHA256(k_i * E_i)$ (ECDH shared secret known only by $N_r$ and $N_i$)
```
Markdown rendering is broken because of a misplaced \`, we should probably remove all the \` when we already have \$.

Reply:
Wow, I didn't even know that worked.  But if we take out ` the formatting changes from monospaced to normal, which would be confusing unless we do it everywhere.

Fixed for now...